qcog_python_client/qcog/pytorch/handler.py

```diff
@@ -0,0 +1,131 @@
+"""Defines the interface for the chain of responsibility pattern.
+
+Commands get dispatched through the chain of handlers. Each handler
+checks if it can handle the command. If it can, it executes the command.
+If the handler returns a new command, it gets dispatched again though the chain.
+
+Each handler implements an automatic retry mechanism.
+The number of attempts and the time to wait can be configured for each handler.
+
+In case of error, the `revert` method is called to undo any changes made by the handler.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import enum
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Coroutine, Generic, TypeAlias, TypeVar
+
+from pydantic import BaseModel
+
+
+class BoundedCommand(BaseModel):
+    """Command type."""
+
+    command: Command
+    dispatch_next: bool = False
+
+
+CommandPayloadType = TypeVar("CommandPayloadType", bound=BoundedCommand)
+ToolName: TypeAlias = str
+ToolFn: TypeAlias = Callable[..., Coroutine[Any, Any, dict]]
+
+
+class Command(enum.Enum):
+    """Command enum."""
+
+    save_parameters = "save_parameters"
+    discover = "discover"
+    validate = "validate"
+    upload = "upload"
+    revert_all = "revert_all"
+    train = "train"
+
+
+class Handler(ABC, Generic[CommandPayloadType]):
+    """Interface for the chain of responsibility pattern."""
+
+    head: Handler  # Reference to the head of the chain
+    next: Handler | None = None
+    attempts: int = 3
+    retry_after: int = 3
+    commands: tuple[Command]
+    get_tool: Callable[[ToolName], ToolFn]
+    context: dict
+
+    @abstractmethod
+    async def handle(self, payload: CommandPayloadType) -> CommandPayloadType | None:
+        """Handle the data."""
+        ...
+
+    @abstractmethod
+    async def revert(self) -> None:
+        """Revert the changes."""
+        ...
+
+    def set_next(self, next_component: Handler, head: Handler) -> Handler:
+        """Set the next component."""
+        self.next = next_component
+        self.head = head
+        return next_component
+
+    async def dispatch(self, payload: CommandPayloadType) -> Handler:
+        """Dispatch the payload through the chain."""
+        # If the command is a `revert_all` command
+        # and the handler has already been executed
+        # we want to revert the state of the handler
+        # and pass the command to the next handler
+        if payload.command == Command.revert_all:
+            await self.revert()
+            if self.next:
+                return await self.next.dispatch(payload)
+            return self
+
+        # If the handler matches the command in the payload
+        # Attempt the execution of the command
+        if payload.command in self.commands:
+            for i in range(self.attempts):
```

Comment:
so I don't think that this is going to do what you want. If we have 3 attempts then on the first attempt we call `execute_and_dispatch_next` and it fails

That drops us into the except block and then we call revert (let us assume that succeeds). Then we sleep. Then we call `execute_and_dispatch_next` again.

On the second call let's say it fails. Now since that attempt is inside the except block and if it fails it will throw so we never attempt more than 2 times.

What you want to do is not call the 2nd `execute_and_dispatch_next` and then use an else block for the for loop to catch if we've tried all attempts and not returned https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

Reply:
Doh! you are right! I deserve punishment for this